### PR TITLE
add user module

### PR DESCRIPTION
**Pull request by Templada**

### CODE REVIEW

File: src/reuse/modules/util/userSettings.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+export class UserSettings {
+  private vlf = new VerboseLoggerFactory("util", "user");
+  private _srvInstance = null;
+
+  private async _initForUserSetting(user: string, password: string): Promise<any> {
+    const vl = this.vlf.initLog(this._initForUserSetting);
+    //TODO: Add vl.log msg
+    const params = browser.config.params;
+    if (params?.systemUrl) {
+      this._srvInstance = await service.odata.init(`${params.systemUrl}/sap/opu/odata/UI2/INTEROP`, user, password);
```
Comment (Templada) [inline src/reuse/modules/util/userSettings.ts:19]: browser.config.params.odataUrl can not be taken since the default for odataUrl is
`/sap/opu/odata/sap` and the `INTEROP`service url is `/sap/opu/odata/UI2/INTEROP/` so the automatic concat of `/sap/opu/odata/sap/UI2/INTEROP`would lead to an incorrect URL

File: test/reuse/util/userSettings/public.key [hunk 2]
```diff
@@ -0,0 +1,8 @@
+-----BEGIN RSA PUBLIC KEY-----
```
Comment (Templada) [inline test/reuse/util/userSettings/public.key:1]: @DIFSRIP | @cvakiitho : do we need to set/overwrite the priv key in the gh-actions

File: test/reuse/util/userSettings/test.user.conf.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const path = require("path");
+const merge = require("deepmerge");
+const qmateConfiguration = require("../../../helper/configurations/chrome.headless.conf");
+const data = require("./data/data.local.json");
+exports.config = merge(qmateConfiguration.config, {
+  maxInstances: 1,
+  params: {
+    systemUrl: data.systemUrl,
```
Comment (Templada) [inline test/reuse/util/userSettings/test.user.conf.js:8]: @DIFSRIP: Lets dicsuss the data handling

File: src/reuse/modules/util/formatter.ts [hunk 4]
```diff
@@ -143,7 +143,7 @@ export class Formatter {
    * const formattedDate = util.formatter.formatDate(date, "mmm dd, yyyy");
    * // returns "Apr 03, 2022"
    */
-  formatDate(date: Date, format: DateFormatsType, locale = "en-US"): string | Date {
+  formatDate(date: Date, format: DateFormatsType, locale = "en-US"): string | Date { //TODO: suggestion to add process.env.USER_SETTINGS_DATE_FORMAT as a default value for format
```
Comment (Templada) [inline src/reuse/modules/util/formatter.ts:146]: let's discuss this (comment)

File: src/reuse/modules/common/date.ts [hunk 5]
```diff
@@ -23,7 +23,7 @@ export class DateModule {
    * @returns {String} The date in the given format.
    * @example const date = await common.date.getToday("mm/dd/yyyy");
    */
-  getToday(format: DateFormatsType = DateFormats.OBJECT): Date | string {
+  getToday(format: DateFormatsType = process.env.USER_SETTINGS_DATE_FORMAT as DateFormats.OBJECT): Date | string {
```
Comment (vlad9719) [inline src/reuse/modules/common/date.ts:26]: Is it possible that process.env.USER_SETTINGS_DATE_FORMAT is not defined here, for instance if no login happens? If so, I would add a fallback default option, here and in the other places in this module
```suggestion
  getToday(format: DateFormatsType = process.env.USER_SETTINGS_DATE_FORMAT ?  process.env.USER_SETTINGS_DATE_FORMAT as DateFormats.OBJECT : DateFormats.OBJECT): Date | string {
```

File: src/reuse/modules/util/constants/formatter.constants.ts [hunk 6]
```diff
@@ -8,5 +8,17 @@ export enum DateFormats {
   MONTH_DAY_YEAR_COMMA = "mmm dd, yyyy",
   MONTH_DAY_YEAR_COMMA_SHORT = "mmm d, yyyy",
   DATETIME = "datetime",
-  OBJECT = "object"
-}
+  OBJECT = "object",
+  GREGORIAN_DOT = "dd.mm.yyyy (Gregorian Date)",
```
Comment (vlad9719) [inline src/reuse/modules/util/constants/formatter.constants.ts:12]: Is it necessary to add those brackets with comment in the end, here and in the constants below?

File: test/reuse/util/userSettings/test.user.conf.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+const path = require("path");
+const merge = require("deepmerge");
+const qmateConfiguration = require("../../../helper/configurations/chrome.headless.conf");
+exports.config = merge(qmateConfiguration.config, {
+  params: {
+    import: {
+      data: "./data/"
+    },
+    // applyUserSettingsForS4: true
```
Comment (vlad9719) [inline test/reuse/util/userSettings/test.user.conf.js:9]: Do you need this param in the config? If not, consider removing it

File: src/reuse/modules/util/formatter.ts [hunk 8]
```diff
@@ -156,64 +156,44 @@ export class Formatter {
...
+
+    sec = sec.toString().padStart(2, "0");
+    min = min.toString().padStart(2, "0");
+    hour = hour.toString().padStart(2, "0");
+    dd = (dd < 10 && format !== DateFormats.MONTH_DAY_YEAR_COMMA_SHORT) ? dd.toString().padStart(2, "0") : dd;  
+    mm = mm.toString().padStart(2, "0");
+    yy = yy.toString().padStart(2, "0");
+
+
+    const formatMappings = {
```
Comment (vlad9719) [inline src/reuse/modules/util/formatter.ts:169]: I like this constant, but I would keep it outside of the class, just to make the function a bit shorter and easier to read?

File: src/reuse/modules/util/formatter.ts [hunk 9]
```diff
@@ -156,64 +156,44 @@ export class Formatter {
...
-    }
-
-    if (dd < 10 && format !== DateFormats.MONTH_DAY_YEAR_COMMA_SHORT) {
-      dd = `0${dd}`;
+    let yy: number | string = yyyy % 100;
+
+    sec = sec.toString().padStart(2, "0");
+    min = min.toString().padStart(2, "0");
+    hour = hour.toString().padStart(2, "0");
+    dd = (dd < 10 && format !== DateFormats.MONTH_DAY_YEAR_COMMA_SHORT) ? dd.toString().padStart(2, "0") : dd;
```
Comment (vlad9719) [inline src/reuse/modules/util/formatter.ts:164]: I would move the part of the expression before '?' into a separate local boolean function, `shouldPadDay`. Thus readability could be improved

**Review by vlad9719 — CHANGES_REQUESTED** [review]
Thank you for your work!

Looks good overall, but I've added a couple of suggestions, and also questions, where I did not understood the code. Please have a look at those

File: src/reuse/modules/util/userSettings.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    if (!this._srvInstance) {
+      await this._initForUserSetting(user, password);
+    }
+    const res = await service.odata.get(this._srvInstance, "UserProfileProperties", { id: "NUMBER_FORMAT", shellType: "FLP" });
+    const resUserData = await service.odata.get(this._srvInstance, "UserProfilePropertyValues", { id: "NUMBER_FORMAT", shellType: "FLP", value: res.value });
+    process.env.USER_SETTINGS_NUMBER_FORMAT = resUserData.description;
+    util.console.info(`Number Format: ${process.env.USER_SETTINGS_NUMBER_FORMAT} was set.`);
+  }
+
+  public async apply(user: string, password: string, isS4: boolean) {
```
[the remark this answers is not in the document]
Comment (Templada) [reply src/reuse/modules/util/userSettings.ts:77]: will be browser.config.params.applyS4UserSettings or so then in the end

File: test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution & Verification: Set User Date", async function () {
+      await util.userSettings.setDateFormatFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
+      const date = await common.date.getToday("yyyy/mm/dd");
+      const userDate = await common.date.getToday(process.env.USER_SETTINGS_DATE_FORMAT);
+      //Ensure returned date matches 10 digits
+      common.assertion.expectEqual(date.length, 10);
```
[the remark this answers is not in the document]
Comment (Templada) [reply test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js:15]: got it, i can replace it with a meaningful constant - it is 10 digits because of the given date formated passed in to the getToday (yyyy/mm/dd), so it should return sth like 2024/08/02 which is 10 chars long.

File: src/reuse/modules/common/date.ts (hunk 5 above)
Comment (Templada) [reply src/reuse/modules/common/date.ts:26]: i think that should already be the case, but would be more clear with the ternary expression

File: src/reuse/modules/util/constants/formatter.constants.ts (hunk 6 above)
Comment (Templada) [reply src/reuse/modules/util/constants/formatter.constants.ts:12]: before it was, as the response of the odata request is with the brackets, since i cut it now before brackets with replace(), we can probably also delete it from the formatter.constants.ts

File: src/reuse/modules/util/formatter.ts (hunk 8 above)
Comment (Templada) [reply src/reuse/modules/util/formatter.ts:169]: actually i didnt changed that at all, think it is showing as new code as it got slightly moved?

File: src/reuse/modules/util/userSettings.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    } else {
+      throw new Error("System URL is missing in the config file.");
+    }
+  }
+
+  public async setLanguageFromUserSettings(user: string, password: string) {
+    const vl = this.vlf.initLog(this.setLanguageFromUserSettings);
+    if (!this._srvInstance) {
+      await this._initForUserSetting(user, password);
+    }
```
[the remark this answers is not in the document]
Comment (Templada) [reply src/reuse/modules/util/userSettings.ts:28]: moved it to the init and called in every method the init

File: test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution & Verification: Set User Date", async function () {
+      await util.userSettings.setDateFormatFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
+      const date = await common.date.getToday("yyyy/mm/dd");
+      const userDate = await common.date.getToday(process.env.USER_SETTINGS_DATE_FORMAT);
+      //Ensure returned date matches 10 digits
+      common.assertion.expectEqual(date.length, DATE_FORMAT_LENGTH);
```
Comment (Templada) [inline test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js:17]: expect type of enum dateformat object instead of length

File: src/reuse/modules/util/userSettings.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (vlad9719) [reply src/reuse/modules/util/userSettings.ts:28]: have you pushed that change? I still see it as before

File: src/reuse/modules/util/userSettings.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    if (browser.config.params?.applyUserSettingsForS4) {
+      try {
+        await this._initForUserSetting(user, password);
+        await this.setDateFormatFromUserSettings(user, password);
+        await this.setLanguageFromUserSettings(user, password);
+        await this.setNumberFormatFromUserSettings(user, password);
+        await this.setTimeFormatFromUserSettings(user, password);
+        await this.setTimeZoneFromUserSettings(user, password);
+      } catch (error) {
+      vl.log(`Function: 'setUserSettingsForS4' failed: Unable to set the UserSettings: ${error}`);
```
[the remark this answers is not in the document]
Comment (vlad9719) [reply src/reuse/modules/util/userSettings.ts:88]: so basically this error would be thrown if someone sets userSettings in the browser.config.params, but they could not be set? Then I would suggest write the message just as `Failed to set the UserSettings for s4: ${error}`

File: src/reuse/modules/util/formatter.ts (hunk 9 above)
Comment (DIFSRIP) [reply src/reuse/modules/util/formatter.ts:164]: but the code is very different, can you please double check

File: src/reuse/modules/util/formatter.ts (hunk 8 above)
Comment (DIFSRIP) [reply src/reuse/modules/util/formatter.ts:169]: can we move it out nevertheless as Vlad suggested?

File: src/reuse/modules/util/userSettings.ts [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+import { VerboseLoggerFactory } from "../../helper/verboseLogger";
+
+/**
+ * @class userSettings
+ * @memberof util
+ */
+
+export class UserSettings {
+  private vlf = new VerboseLoggerFactory("util", "user");
```
Comment (DIFSRIP) [inline src/reuse/modules/util/userSettings.ts:11]:
```suggestion
  private vlf = new VerboseLoggerFactory("util", "userSettings");
```

File: src/reuse/modules/util/userSettings.ts [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+      } else {
+        throw new Error("System URL is missing in the config file.");
+      }
+    }
+  }
+
+  public async setLanguageFromUserSettings(user: string, password: string) {
+    const vl = this.vlf.initLog(this.setLanguageFromUserSettings);
+    await this.initS4UserSettingService(user, password);
+    const res = await service.odata.get(this._srvInstance, "UserProfileProperties", { id: "LANGUAGE", shellType: "FLP" });
```
Comment (DIFSRIP) [inline src/reuse/modules/util/userSettings.ts:37]: I would try catch all the API calls as you did for the init.

File: src/reuse/modules/util/userSettings.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (DIFSRIP) [reply src/reuse/modules/util/userSettings.ts:87]: in general, the logger should not be used for errors. Errors will not be shown in regular mode like that. It is used to log additional info on the function behaviour if executed in verbosity mode. You should throw a regular error

File: test/reuse/util/userSettings/getDateFormatFromUserSettings.spec.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution & Verification: Set User Date", async function () {
+      const userDateFormat = await util.userSettings.getDateFormatFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
+      const date = await common.date.getToday("yyyy/mm/dd");
+      const userDate = await common.date.getToday(userDateFormat);
+      //Ensure returned date matches 10 digits
+      common.assertion.expectEqual(date.length, DATE_FORMAT_LENGTH);
```
Comment (DIFSRIP) [inline test/reuse/util/userSettings/getDateFormatFromUserSettings.spec.js:17]: instead of checking for the length (which might not be working for specific other configs), I would check if both dates are of the same type (DateFormatsType) since the userDateFormat could be different from time to time while the "yyyy/mm/dd" is static

File: test/reuse/util/userSettings/getTimeFormatFromUserSettings.spec.js [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+
+describe("userSettings", function () {
+  describe("getTimeFormatFromUserSettings.spec", function () {
+
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution", async function () {
+      await util.userSettings.getTimeFormatFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
```
Comment (DIFSRIP) [inline test/reuse/util/userSettings/getTimeFormatFromUserSettings.spec.js:12]: no assertion here?

File: test/reuse/util/userSettings/getTimeZoneFromUserSettings.spec.js [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+
+describe("userSettings", function () {
+  describe("getTimeZoneFromUserSettings.spec", function () {
+
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution", async function () {
+      await util.userSettings.getTimeZoneFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
```
Comment (DIFSRIP) [inline test/reuse/util/userSettings/getTimeZoneFromUserSettings.spec.js:11]: no assertion here?

File: test/reuse/util/userSettings/public.key (hunk 2 above)
Comment (DIFSRIP) [reply test/reuse/util/userSettings/public.key:1]: no, you should use the default keys. I guess it is not required to add the key here, can you check?

File: test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js (hunk 13 above)
Comment (DIFSRIP) [reply test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js:17]: same as above as discussed yesterday

File: test/reuse/util/userSettings/setLanguageFromUserSettings.spec.js [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+describe("userSettings", function () {
+  describe("setLanguageFromUserSettings.spec", function () {
+
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution & Verification", async function () {
+      await util.userSettings.setLanguageFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
+      common.assertion.expectEqual(process.env.USER_SETTINGS_LANG_KEY.length, 2);
```
[the remark this answers is not in the document]
Comment (DIFSRIP) [reply test/reuse/util/userSettings/setLanguageFromUserSettings.spec.js:12]: I don't understand what is being tested here

File: test/reuse/util/userSettings/setS4UserSettings.spec.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+    describe("setS4UserSettings.spec - apply all s4 user settings via config<->login", function () {
+      it("Preparation: Set systemUrl and baseUrl ", async function () {
+        browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+        browser.config.baseUrl = util.data.decrypt(data.baseUrl);
+      });
+
+      it("Execution & Verification", async function () {
+        await ui5.navigation.navigateToApplication("BusinessUser-maintain");
+        await ui5.session.login(util.data.decrypt(data.username), util.data.decrypt(data.password));
+        common.assertion.expectEqual(DATE_FORMAT_LENGTH, process.env.USER_SETTINGS_DATE_FORMAT.length);
```
Comment (DIFSRIP) [inline test/reuse/util/userSettings/setS4UserSettings.spec.js:25]: is it possible to add an actual assertion since this line is unrelated to the one before

**Review by DIFSRIP — CHANGES_REQUESTED** [review]
Please see my comments

File: src/reuse/modules/util/userSettings.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Templada) [reply src/reuse/modules/util/userSettings.ts:87]: information of which class is missing then

File: .gitignore [hunk 22]
```diff
@@ -36,7 +36,7 @@ Dictionaries/
 GPUCache/
 exthost Crash Reports/
 Local Storage/
-User/
+[U]ser/
```
[the remark this answers is not in the document]
Comment (Templada) [reply .gitignore:39]: idk where this was from tbh, will remove it

File: .prettierrc.json [hunk 23]
```diff
@@ -1,6 +1,6 @@
 {
     "tabWidth": 2,
-    "printWidth": 300,
+    "printWidth": 200,
```
[the remark this answers is not in the document]
Comment (Templada) [reply .prettierrc.json]: reverted

File: test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js (hunk 13 above)
Comment (Templada) [reply test/reuse/util/userSettings/setDateFormatFromUserSettings.spec.js]: as discussed js can not handle types as we expected in our meeting, therefore keeping the date evaluation only.

File: src/reuse/modules/util/formatter.ts (hunk 9 above)
Comment (Templada) [reply src/reuse/modules/util/formatter.ts]: took it from main, as discussed we will refactor it with the suggested code in another PR

File: src/reuse/modules/util/formatter.ts (hunk 8 above)
Comment (Templada) [reply src/reuse/modules/util/formatter.ts]: as said we will do that in another PR

File: test/reuse/util/userSettings/setS4UserSettings.spec.js (hunk 21 above)
Comment (Templada) [reply test/reuse/util/userSettings/setS4UserSettings.spec.js]: added expectDefined() as it is to generic to have test against static values.

Comment (Templada) [reply test/reuse/util/userSettings/setS4UserSettings.spec.js]: regarding the Language Key comment: "I don't understand what is being tested here" -> The Language Key can only be 2 chars long therefore checking if it matches a length of 2

**Comment by Templada** [comment]
@vlad9719: Can you please check and review again, everything should be resolved now.

Private key still don't work - agreed with Benny that we comment it out after successful run, since it is connecting to an internal system and is likely to fail due to connectivity issues (maint/downtimes/etc).

So if you want to check if the tests are fine/running green, please execute them locally.
All tests should run w/ success:
image
image

File: test/reuse/util/userSettings/setNumberFormatFromUserSettings.spec.js [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+
+describe("userSettings", function () {
+  describe("setNumberFormatFromUserSettings.spec", function () {
+    
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution", async function () {
+      await util.userSettings.setNumberFormatFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
```
[the remark this answers is not in the document]
Comment (Templada) [reply test/reuse/util/userSettings/setNumberFormatFromUserSettings.spec.js:12]: done

File: test/reuse/util/userSettings/setTimeFormatFromUserSettings.spec.js [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+
+describe("userSettings", function () {
+  describe("setTimeFormatFromUserSettings.spec", function () {
+
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution", async function () {
+      await util.userSettings.setTimeFormatFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
```
[the remark this answers is not in the document]
Comment (Templada) [reply test/reuse/util/userSettings/setTimeFormatFromUserSettings.spec.js:12]: done

File: test/reuse/util/userSettings/setTimeZoneFromUserSettings.spec.js [hunk 26]
```diff
@@ -0,0 +1,15 @@
...
+
+describe("userSettings", function () {
+  describe("setTimeZoneFromUserSettings.spec", function () {
+
+    it("Preparation: Set systemUrl ", async function () {
+      browser.config.params.systemUrl = util.data.decrypt(data.systemUrl);
+    });
+
+    it("Execution", async function () {
+      await util.userSettings.setTimeZoneFromUserSettings(util.data.decrypt(data.username), util.data.decrypt(data.password));
```
[the remark this answers is not in the document]
Comment (Templada) [reply test/reuse/util/userSettings/setTimeZoneFromUserSettings.spec.js:11]: done

File: test/reuse/util/userSettings/getTimeFormatFromUserSettings.spec.js (hunk 18 above)
Comment (Templada) [reply test/reuse/util/userSettings/getTimeFormatFromUserSettings.spec.js]: added assertion

File: test/reuse/util/userSettings/setLanguageFromUserSettings.spec.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (Templada) [reply test/reuse/util/userSettings/setLanguageFromUserSettings.spec.js]: @vlad9719 : no since it could be changed on system, therefore only length of language key can be checked

File: test/reuse/util/userSettings/getDateFormatFromUserSettings.spec.js (hunk 17 above)
Comment (Templada) [reply test/reuse/util/userSettings/getDateFormatFromUserSettings.spec.js]: as discussed in the meeting, type cant be checked for now as we dont use ts, therefore deleted the length check and kept the date verification

File: src/reuse/modules/util/userSettings.ts (hunk 16 above)
Comment (Templada) [reply src/reuse/modules/util/userSettings.ts]: move all api calls in _getXXXXresponse functions and added try catch. Error handling will be improved later on in different PR as we agreed

**Review by vlad9719 — APPROVED** [review]
Looks good!
